File: course-material/labs/inheritance/coding-exercises/solution/ast/rpn-builder.cpp
#include "rpn-builder.h"


RPNBuilder& RPNBuilder::literal(double x)
{
    this->_stack.push(std::make_shared<const Literal>(x));

    return *this;
}

RPNBuilder& RPNBuilder::variable(const std::string& identifier)
{
    this->_stack.push(std::make_shared<const Variable>(identifier));

    return *this;
}

RPNBuilder& RPNBuilder::add()
{
    auto x = this->_stack.top();
    this->_stack.pop();
    auto y = this->_stack.top();
    this->_stack.pop();

    std::vector<std::shared_ptr<const Expression>> operands{ x, y };

    this->_stack.push(std::make_shared<const Addition>(operands));

    return *this;
}

RPNBuilder& RPNBuilder::multiply()
{
    auto x = this->_stack.top();
    this->_stack.pop();
    auto y = this->_stack.top();
    this->_stack.pop();

    std::vector<std::shared_ptr<const Expression>> operands{ x, y };

    this->_stack.push(std::make_shared<const Multiplication>(operands));

    return *this;
}

std::shared_ptr<const Expression> RPNBuilder::build() const
{
    return this->_stack.top();
}

File: course-material/labs/inheritance/coding-exercises/student/ast/java/ast/RPNBuilder.java
package ast;

import java.util.ArrayList;
import java.util.Stack;

public class RPNBuilder {
    private Stack<Expression> stack;

    public RPNBuilder() {
        stack = new Stack<Expression>();
    }

    public RPNBuilder add() {
        Expression a = stack.pop();
        Expression b = stack.pop();

        ArrayList<Expression> operands = new ArrayList<>();
        operands.add(a);
        operands.add(b);

        stack.push(new Addition(operands));

        return this;
    }

    public RPNBuilder multiply() {
        Expression a = stack.pop();
        Expression b = stack.pop();

        ArrayList<Expression> operands = new ArrayList<>();
        operands.add(a);
        operands.add(b);

        stack.push(new Multiplication(operands));

        return this;
    }

    public RPNBuilder literal(double x) {
        stack.push(new Literal(x));

        return this;
    }

    public RPNBuilder variable(String identifier) {
        stack.push(new Variable(identifier));

        return this;
    }

    public Expression build() {
        return stack.peek();
    }
}
